### ISSUE TITLE
Distribution of the Pepper Flash player is very dubious legally...

### ISSUE THREAD

**Issue by spl237** [Botspot/scratch2#2]
Just a friendly warning, having noticed this repo...

The Adobe Flash player was removed from RPiOS for a reason, and Adobe were very clear that they wanted it completely removed, with no trace remaining; we had to purge all traces of it from GitHub histories, for example. It is regarded as a major security risk - which is why it was withdrawn - and Adobe may take action against anyone found to be distributing it without their permission (which they would definitely not give...)

My strong recommendation would be to take this repo down, but if you continue to host it, be aware that it is at your own risk and may result in undesired attention from Adobe's legal department.

**Comment by Botspot** [comment]
Thank you for bringing this to my attention. We will discuss this internally and see if the chance of legal backlash outweighs the 2000+ people currently enjoying continued use of Scratch2.

On a similar note, @spl237, do you happen to know how RPT legally navigated the hosting of proprietary ChromeOS DRM binaries within the `libwidevinecdm0` package? I think the answer to that could shed light on how my team can proceed here.

**Comment by spl237** [comment]
> Thank you for bringing this to my attention. We will discuss this internally and see if the chance of legal backlash outweighs the 2000+ people currently enjoying continued use of Scratch2.

Seriously, even without the legal issues, the continued use of Flash Player is extremely unwise for security reasons - Adobe were pushing patches for it on a monthly basis to fix the security holes; the reason it was withdrawn is that they realised it was a battle they no longer wanted to fight. 

I have no idea how old the version you are hosting is, but even the final release has now had no patches for a couple of years, which will mean it has multiple unpatched exploits; older versions would be even more risky. I wouldn't want it on any system of mine under any circumstances.

> On a similar note, @spl237, do you happen to know how RPT legally navigated the hosting of proprietary ChromeOS DRM binaries within the `libwidevinecdm0` package? I think the answer to that could shed light on how my team can proceed here.

We negotiated with Google. 

I can tell you for absolute certain that whatever you say to Adobe, that won't work here - they will absolutely not agree to you (or anyone else) distributing it; they want Flash completely removed from the net.

**Comment by Botspot** [comment]
In practice, with scratch2 being an offline app, I am led to believe that any attack surface would be negligible.
@spl237, I would completely agree with your concern if this application had cloud or online features, but to the best of my knowledge it is completely offline without any telemetry or internet activity.

To help prevent anybody from trying to use libpepflashplayer.so for other things, my team and I decided that it is best to remove it from this repository. While it is currently accessible in the git history, I plan to force-push a squash commit to remove this file from the git history. (As soon as I figure out how to do that)

I did request advice from my team about this whole situation, and one of them directed me to the [Chrome terms of service](https://web.archive.org/web/20200421061947/https://www.google.com/intl/en/chrome/terms/). To the best of our understanding, this repository has been in compliance all along, (or else the official scratch2 package was in violation the whole time), and with these recent changes we believe the repository now goes above and beyond Adobe's requirements.

**Comment by spl237** [comment]
> I did request advice from my team about this whole situation, and one of them directed me to the [Chrome terms of service](https://web.archive.org/web/20200421061947/https://www.google.com/intl/en/chrome/terms/). To the best of our understanding, this repository has been in compliance all along, (or else the official scratch2 package was in violation the whole time), and with these recent changes we believe the repository now goes above and beyond Adobe's requirements.

I'm sorry, but you are mistaken on this.

The Pepper Flash library was never distributed as part of Chrome / Chromium, certainly not on Pi, and the distribution terms for Chrome do not apply to it. (It was never in the scratch2 package, either, so that's not relevant to this discussion.)

It was an entirely separate library which was distributed by us in our chromium-mods package with permission from Adobe. It is not part of Chrome or Scratch, it is a closed-source software library under a commercial license, which may not be distributed by anyone without express permission from Adobe.

We had permission to distribute the Pepper Flash library. Our permission to distribute was withdrawn by Adobe when they retired Flash, at which point we removed the library from all repos, packages and even old images. 

Unless you have such an agreement - and you do not, because Adobe will no longer agree to anyone distributing the library under any circumstances - you are violating Adobe's copyright by hosting or distributing the library. 

Scratch 2 will not work without the Flash library. The Flash library may no longer be distributed by anyone. Therefore, Scratch 2 can no longer be installed on a Pi - and in fact, even on any old images, as soon as they are updated from apt, the library will be removed and Scratch 2 will stop working. 

Users can still use Scratch 1 if their Pi will not run Scratch 3, but Scratch 2 can no longer legally be installed on a Pi. I understand why you would like to work around this, but it is simply not possible for anyone to legally host or otherwise distribute the Flash library which is required to do so any more.

(Further, the way the library is installed so that it can be used by an Electron instance means that, if I understand correctly, it will also be available to Chromium running on the same Pi, so it will render Chromium on that Pi vulnerable to attack whenever it is used - the offline nature of Scratch isn't really the issue from a security point of view. But that is a secondary concern compared to the legalities.)

To my knowledge there is no legal way to do what you are trying to do here. If you remove the library, Scratch will not work, and you have no choice but to remove the library.

**Comment by theofficialgman** [comment]
> I'm sorry, but you are mistaken on this.
> 
> The Pepper Flash library was never distributed as part of Chrome / Chromium, certainly not on Pi, and the distribution terms for Chrome do not apply to it.

no, you are mistaken. it was distributed as part of chrome (x86, and x86_64) and well on chromeOS (x86, x86_64, and arm). in addition, let me remind you of when a user clearly pointed out this is indeed pepperflash and not adobe's 
open source flashplayer library, even if it was adobe that supplied it to you.  https://forums.raspberrypi.com/viewtopic.php?t=162525&start=50#p1055655

we are not subject to your license with adobe, our `libpepperflashplayer.so` binary was sourced from an arm chromeOS recovery image and as such is subject to the chrome license agreement that Botspot linked above. the binaries found in the chromeOS recovery images are a binary match for the ones you supplied in raspbian images (when you match the versions) so they are the same file even.

> Further, the way the library is installed so that it can be used by an Electron instance means that, if I understand correctly, it will also be available to Chromium running on the same Pi, so it will render Chromium on that Pi vulnerable to attack whenever it is used - the offline nature of Scratch isn't really the issue from a security point of view. But that is a secondary concern compared to the legalities.

this is not the case anymore, the javascript of scratch2 has been modified and the binary location of `libpepperflashplayer.so` and electron changed to reside within the scratch2 folder. it is entirely separated and offline.

**Comment by spl237** [comment]
> we are not subject to your license with adobe, our `libpepperflashplayer.so` binary was sourced from an arm chromeOS recovery image 

The comment you removed a couple of weeks ago, but which is still in your GitHub history, directly contradicts this statement, which says you sourced it from one of our images - https://github.com/Botspot/scratch2/commit/db18ffd5842da373e0d208d74f00bb2079feb111

Our binaries were shipped directly to us from Adobe. You do not have any license to distribute them.

**Comment by theofficialgman** [comment]
> > we are not subject to your license with adobe, our `libpepperflashplayer.so` binary was sourced from an arm chromeOS recovery image
> 
> The comment you removed a couple of weeks ago, but which is still in your GitHub history, directly contradicts this statement, which says you sourced it from one of our images - [db18ffd](https://github.com/Botspot/scratch2/commit/db18ffd5842da373e0d208d74f00bb2079feb111)

please read my full response before responding 😄 

>  the binaries found in the chromeOS recovery images are a binary match for the ones you supplied in raspbian images (when you match the versions) so they are the same file even.

regardless if Botspots previous hosting was sourced from the raspbian images, I have verified and sourced the binaries from a chromeOS recovery image and they are a binary match. so even the previous binaries ARE sourced from a chromeOS recovery image. as such, it is not subject to your license

**Comment by theofficialgman** [comment]
also @Botspot I'll gladly cleanup the repo history if you give me push access

**Comment by Botspot** [comment]
Github repo history has been removed, as you requested earlier, @spl237.

Also, to be as far in the clear as possible, I went ahead and swapped out the libpepflashplayer.so with the very last available version from Google's [ChromeOS recovery image v13505.63.0](https://dl.google.com/dl/edgedl/chromeos/recovery/chromeos_13505.63.0_elm_recovery_stable-channel_mp-v2.bin.zip).
As such, this fully dismisses the argument that I am violating RPT's contract with Google, as files involved with the agreement are no longer used.

Going forward, only Google or Adobe have legal capacity to object to this distribution of Scratch2. And if either of them object, I will kindly refer them to their own [TOS](https://web.archive.org/web/20200421061947/https://www.google.com/intl/en/chrome/terms/).

At this point, I believe that everything has been sorted out.
Thanks again to @spl237 for taking the time to bring the issue up. Thank you for being patient with us!
